tests/test_client.py

```diff
@@ -0,0 +1,184 @@
+import pytest
+import testinfra
+import subprocess
+import os
+import time
+import json
+
+
+@pytest.fixture
+def entities():
+    return {
+        'organization_label': 'Test_Organization',
+        'activation_key': 'Test AK',
+        'product': 'Test Product',
+        'product_label': 'Test_Product',
+        'yum_repository': 'Zoo',
+        'yum_repository_label': 'Zoo',
+        'errata_id': 'RHEA-2012:0055',
+        'container_repository_label': 'foremanbusybox'
+    }
+
+@pytest.fixture(scope='session')
+def client(request):
+    name = "smoker-client-%s" % request.param.replace('/', '_').replace(':', '_')
+    podman_id = subprocess.check_output(['podman', 'ps', '-q', '--filter', "name=%s" % name])
+
+    if not podman_id:
+        subprocess.check_output(['podman', 'run', '--name', name, '-dt', request.param]).decode().strip()
+
+    # return a testinfra connection to the container
+    yield testinfra.get_host("podman://" + name)
+
+@pytest.fixture(scope='session')
+def smoker():
+    return testinfra.get_host("local://localhost")
+
+@pytest.fixture
+def server_hostname(variables):
+    assert variables['server_hostname']
+    return variables['server_hostname']
+
+@pytest.fixture
+def foreman_version(variables):
+    assert variables['foreman_version']
+    return variables['foreman_version']
+
+def pytest_generate_tests(metafunc):
+    with open(metafunc.config.option.variables[0]) as file:
+      variables = json.load(file)
+      metafunc.parametrize('client', variables['clients'], indirect=True, scope='session')
+
+def teardown_module():
```

Comment:
Better than a teardown module is to do it as part of the fixture. https://docs.pytest.org/en/stable/fixture.html#fixture-finalization-executing-teardown-code

Reply:
Yea, that was my original implementation. As soon as I introduced the parameterization, that stopped working. Even when I try to specify `session` or `module` based scope the code after the `yield` in the client fixture gets called after every test which destroys the container between tests. I tried many things, perhaps someone else might have better luck. This worked nicely.

Reply:
For what it's worth: for good test isolation, I think you should delete the container between runs.

I noticed that current Fedora and Debian don't include systemd in their container and it wouldn't surprise me if a future CentOS wouldn't do the same thing. Perhaps we need to prepare some Dockerfiles and build our own base images (locally) which also include `subscription-manager`. That way you have a guaranteed fresh container without any state. You could launch the container with `--rm` to make sure it's gone after it stops.

Reply:
I see it as trying to walk that line between building expensive, fully isolated test cases vs. orchestrating and asserting across a setup to reduce the time to execute. Seems we can:

 a) have small dedicated tests with 1-2 assertions but are intentionally order dependent (similar to bats today)
 b) have fewer tests but multiple assertions within said test to group testing concerns but reduce repeated setup
 c) have fully isolated, dedicated tests with setup and teardown that do not aim to minimize run time or re-use

I believe that a/b are similar to the acceptance tests in the Puppet modules in style.

Reply:
> I believe that a/b are similar to the acceptance tests in the Puppet modules in style.

I think rspec mostly follows the order as well but our contexts aren't really dependent on them.

What mostly matters is what happens when you use python-xdist. Looks like now we're not using it, but if you run them with `pytest -n 4`. pytest is more aggressive in assuming tests are independent.

There is https://smarie.github.io/python-pytest-steps/ but I have no experience with it and it also states it's untested with python-xdist.

I would prefer to use option c with well chosen fixtures to minimize run time. For example, a fixture that gives you a running container where subscription-manager is already present. When using podman, that can cache the image to reduce the time needed.